src/libmongoc/src/mongoc/mongoc-error.c

```diff
@@ -268,9 +270,13 @@ _mongoc_error_is_not_master (bson_error_t *error)
    switch (error->code) {
    case MONGOC_SERVER_ERR_NOTMASTER:
    case MONGOC_SERVER_ERR_NOTMASTERNOSLAVEOK:
+   case MONGOC_SERVER_ERR_LEGACYNOTPRIMARY:
       return true;
-   default:
+      /* All errors where no code was found are marked as MONGOC_ERROR_QUERY_FAILURE */
+   case MONGOC_ERROR_QUERY_FAILURE:
```

Comment:
I dislike this part. `_mongoc_cmd_check_ok` and `_mongoc_cmd_check_ok_no_wce` rely on `_parse_error_reply` to extract message and code information from a `bson_t` reply. If no code is found in the reply document, a code of `0` is returned.

The two check methods then take this information and build a `bson_error_t` of it, but only after changing the error code to `MONGOC_ERROR_QUERY_FAILURE` (17) if it was 0, essentially defaulting to that error code. I've covered for this by checking for that error code here, but I feel like this is a dirty workaround. `_mongoc_cmd_check_ok_no_wce` and `_mongoc_cmd_check_ok` should not change the error code, but instead this logic should be moved to wherever the `MONGOC_ERROR_QUERY_FAILURE` is needed. I'm not very familiar with this part of libmongoc, so if you have any pointers to where this might require changes I'm happy to investigate.

Reply:
Tracked in CDRIVER-3928.